CMakeLists.txt

```diff
@@ -34,26 +37,43 @@ target_link_libraries(vx3_node_daemon PUBLIC ${Boost_LIBRARIES})
 
 
 # CUDA options
-if(NOT DEFINED CUDA_DEBUG)
-  set(CUDA_DEBUG "-G or not" ON)
-  message("Set CUDA_DEBUG to ON.")
-endif(NOT DEFINED CUDA_DEBUG)
+find_package(CUDA) # for CUDA_SELECT_NVCC_ARCH_FLAGS function
+# if one want to force use certain Gpu Architecture, one can add `CUDA_ARCH` like `cmake -DCUDA_ARCH=arch=compute_61,code=sm_61 ..`
+if (NOT DEFINED CUDA_ARCH)
+  CUDA_SELECT_NVCC_ARCH_FLAGS(RET_ARCH_FLAGS)
+  message("CUDA_SELECT_NVCC_ARCH_FLAGS returns: " ${RET_ARCH_FLAGS})
+  list(LENGTH RET_ARCH_FLAGS FLAG_LENGTH)
+  message("Flag length: " ${FLAG_LENGTH})
+  if (FLAG_LENGTH GREATER 3)
+    # CUDA_SELECT_NVCC_ARCH_FLAGS resulted more than 3 items, which is probably GPU not found error.
+    message(FATAL_ERROR "Nvidia GPU not found. Abort.")
+  endif()
+  list(GET RET_ARCH_FLAGS -1 CUDA_ARCH) # use newest architecture
+endif()
+message("Using GPU Architecture: " ${CUDA_ARCH})
+set(ARCH_FLAGS "-gencode" ${CUDA_ARCH})
+list(APPEND CMAKE_CUDA_FLAGS -gencode=${CUDA_ARCH})
+
 set_target_properties(voxelyze.3.0 PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
 set_target_properties(vx3_node_worker PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
 set_target_properties(vx3_test PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
-    #-G means debug info in GPU, so you can step in GPU kernels
-    if (CUDA_DEBUG)
-      message("Compile with debug information.")
-      target_compile_options(voxelyze.3.0     PUBLIC $<$<COMPILE_LANGUAGE:CUDA>:-G -arch=compute_60 -code=compute_60>)
-      target_compile_options(vx3_node_worker  PUBLIC $<$<COMPILE_LANGUAGE:CUDA>:-G -arch=compute_60 -code=compute_60>)
-      target_compile_options(vx3_test         PUBLIC $<$<COMPILE_LANGUAGE:CUDA>:-G -arch=compute_60 -code=compute_60>)
-    else (CUDA_DEBUG)
-      target_compile_options(voxelyze.3.0     PUBLIC $<$<COMPILE_LANGUAGE:CUDA>: -arch=compute_60 -code=compute_60>)
-      target_compile_options(vx3_node_worker  PUBLIC $<$<COMPILE_LANGUAGE:CUDA>: -arch=compute_60 -code=compute_60>)
-      target_compile_options(vx3_test         PUBLIC $<$<COMPILE_LANGUAGE:CUDA>: -arch=compute_60 -code=compute_60>)
-    endif(CUDA_DEBUG)
-list(APPEND CMAKE_CUDA_FLAGS "-gencode=arch=compute_60,code=sm_60")
 
+string(TOUPPER "${CMAKE_BUILD_TYPE}" uppercase_CMAKE_BUILD_TYPE)
+if(uppercase_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
```

Comment:
discussion:

I accidentally made the "Release" build mode by default instead of "Debug", do you guys think this is a good feature? Should we release our code this way?